hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/records/RecordCache.java

```diff
@@ -37,79 +45,117 @@
+     * transactions with the same ID that also came to consensus.
      *
-     * @param transactionID The transaction ID to lookup
-     * @return The {@link TransactionRecord}s, if any, or else an empty list.
-     */
-    @NonNull
-    List<TransactionRecord> getRecords(@NonNull TransactionID transactionID);
-
-    /**
-     * Gets a list of all records for the given {@link AccountID}. The {@link AccountID} is the account of the Payer of
-     * the transaction.
+     * <p>The {@code childRecords} list contains a list of all child transactions of the original user transaction.
+     * Duplicate transactions never have child transactions.
      *
-     * @param accountID The accountID of the Payer of the transactions
-     * @return The {@link TransactionRecord}s, if any, or else an empty list.
+     * @param nodeIds The IDs of every node that submitted a transaction with the txId that came to consensus and was
+     *                handled. This is an unordered set, since deterministic ordering is not required for this in-memory
+     *                data structure
+     * @param records Every {@link TransactionRecord} handled for every user transaction that came to consensus
      */
-    @NonNull
-    List<TransactionRecord> getRecords(@NonNull AccountID accountID);
+    record History(
+            @NonNull Set<Long> nodeIds,
+            @NonNull List<TransactionRecord> records,
+            @NonNull List<TransactionRecord> childRecords) {
 
-    /**
-     * Gets the "primary" receipt for the given {@link TransactionID}. A user may submit multiple transactions with the
-     * same {@link TransactionID} to the network by submitting to multiple nodes. In that case, there will be multiple
-     * receipts. However, only the first of these will be "primary" and all others will be duplicates.
-     *
-     * @param transactionID The transaction ID to lookup
-     * @return The {@link TransactionReceipt}, if any, or else {@code null}. If the transaction is known to exist but
-     * has not yet been handled, the receipt will have a status of
-     * {@link com.hedera.hapi.node.base.ResponseCodeEnum#UNKNOWN}.
-     */
-    @Nullable
-    default TransactionReceipt getReceipt(@NonNull TransactionID transactionID) {
-        final var receipts = getReceipts(transactionID);
-        return receipts.isEmpty() ? null : receipts.get(0);
+        /**
+         * This receipt is returned whenever we know there is a transaction pending (i.e. we have a history for a
+         * transaction ID), but we do not yet have a record for it.
+         */
+        private static final TransactionReceipt PENDING_RECEIPT =
+                TransactionReceipt.newBuilder().status(UNKNOWN).build();
+
+        /**
+         * Create a new {@link History} instance with empty lists.
+         */
+        public History() {
+            this(new HashSet<>(), new ArrayList<>(), new ArrayList<>());
```

Comment:
Probably doesn't matter, but we *could* be a little defensive here if we used a `TreeSet` or `ConcurrentSkipListSet` to impose a sort order on the nodeIds.